cpp/src/copying/concatenate.cu

```diff
@@ -331,26 +332,82 @@ std::unique_ptr<column> concatenate_dispatch::operator()<cudf::struct_view>()
   return cudf::structs::detail::concatenate(views, stream, mr);
 }
 
+namespace {
+
+/**
+ * @brief Verifies that the sum of the sizes of all the columns to be concatenated
+ * will not exceed the max value of size_type, and verifies all column types match
+ *
+ * @param begin Beginning of range of columns to check
+ * @param end End of range of columns to check
+ *
+ * @throws cudf::logic_error if the total length of the concatenated columns would
+ * exceed the max value of size_type
+ *
+ * @throws cudf::logic_error if all of the input column types don't match
+ */
+template <typename ColIter>
+void bounds_and_type_check(ColIter begin, ColIter end)
+{
+  CUDF_EXPECTS(std::all_of(begin,
+                           end,
+                           [expected_type = (*begin).type()](auto const& c) {
+                             return c.type() == expected_type;
+                           }),
+               "Type mismatch in columns to concatenate.");
+
+  // total size of all concatenated rows
+  auto size_iter = thrust::make_transform_iterator(
+    begin, [](column_view const& c) { return static_cast<size_t>(c.size()); });
+  size_t const total_row_count =
+    std::accumulate(size_iter, size_iter + std::distance(begin, end), static_cast<size_t>(0));
+  CUDF_EXPECTS(total_row_count < std::numeric_limits<size_type>::max(),
```

Comment:
Should this be ?
```suggestion
  CUDF_EXPECTS(total_row_count <= std::numeric_limits<size_type>::max(),
```

Reply:
Good eye.  I left it this way to handle offset columns.  The idea is that if you have N string (or list) columns, the length of the concatenated offsets column isn't directly computable without explicitly add knowledge that "this is an offset column".  For example, if you concatenated two string columns of length 5, you would start with 2 offset columns of length 6.  But the length of the _final_ offset column would be (5 + 5 + 1) not (6 + 6).  

So this sort of dodges that issue by making sure that the top level result is always less than max-1 to leave room for the offset.

Reply:
Should we specialize this for offset column?

Reply:
My thinking is no because the point of this is really to catch overflow and not really to optimize for that 1-in-4-billion case :)  @jlowe  any opinions?

Reply:
Actually, now that I think about it, this doesn't matter.   Even if it pretends to account for the addition offset for the child columns, the child columns will already be acting "conservatively".   For example, it will see that it is concatenating say, 3 offset columns of length 10 and think the result will be length 30 (10 + 10 + 10), when it will really be 27 (9 + 9 + 9).    So it will "fail" when it could have succeeded (but it will just be wrong by <= # offset columns to be concatenated) which is still probably close enough.

Reply:
At my previous employer, a popular saying was "nothing is rare at scale."  If it's easy and cheap to check for all the possible cases, I'd rather have all the bases covered.  If this is difficult, I'd rather file the followup issue tracking the fact that there's an unlikely but uncovered case and we can tackle it later.

Reply:
Comment race with @nvdbaranec's comment.  To be clear, I'd much rather have it fail with a false-positive if we're ever-so-close to overflowing but technically won't rather than generate an invalid column that can crash in a mysterious way (or corrupt data!) much later down the line.  I'm personally OK if the approach would fail if we're a handful of bytes away from the end and thought it would overflow.

Now if we were a handful of megabytes away from the end that would be a different story...

Reply:
I agree, failing is better than corrupted data.

Reply:
Thinking about this further :  this also means the original check prior to this PR would not have worked.  It was just checking top level columns so two string columns that added up to exactly INT_MAX rows would have died later on because it tried to create an offset column of INT_MAX+1 rows.